src/Data/Bin/Ordering.agda

```diff
@@ -0,0 +1,791 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Ordering on Bin and certain its properties.
+------------------------------------------------------------------------
+
+{-# OPTIONS --without-K --safe #-}
+
+module Data.Bin.Ordering where
+
+open import Data.Bin.Base
+open import Data.Bin.Properties
+open import Data.Nat as ℕ using (ℕ; z≤n; s≤s)
+open import Data.Nat.Properties as ℕp using (m+n∸m≡n; m+n∸n≡m)
+open import Data.Product using (_,_; proj₁; proj₂)
+open import Data.Sum using (_⊎_; inj₁; inj₂)
+open import Function using (_∘_; _$_; flip; case_of_; _on_)
+open import Level using () renaming (zero to 0ℓ)
+open import Relation.Binary using
+  (Rel; Reflexive; Symmetric; Antisymmetric; Transitive; Irreflexive; Tri; _⇒_;
+   Decidable; _Preserves_⟶_; _Preserves₂_⟶_⟶_; _Respects₂_; IsPreorder;
+   IsPartialOrder; Poset; IsTotalOrder; IsDecTotalOrder; DecTotalOrder;
+   IsStrictTotalOrder; StrictTotalOrder
+  )
+open import Relation.Binary.PropositionalEquality as P using
+  (_≡_; _≢_; refl; sym; trans; subst; subst₂; cong; cong₂; isEquivalence)
+open import Relation.Binary.PropositionalEquality.Core using (≢-sym)
+import Relation.Binary.Reasoning.Base.Triple as InequalityReasoning
+open import Relation.Nullary using (¬_; yes; no)
+open import Relation.Nullary.Negation using (contradiction)
+
+open Bin
+
+-- The ordering relation
+
+infix 4 _<_ _>_ _≤_
+
+data _<_ : Rel Bin 0ℓ  where
```

Comment:
Out of interest why have you defined the strict order first instead of the standard non-strict order? In every other data type we define `_≤_` first.

Reply:
Suppose that it is needed to define (implement) for some domain D the relations
``≈, <, ≤``, where these relations make D a StrictTotalOrder for ``(≈, <)``,
and TotalOrder for ``(≈, ≤)``.
One can
(A) implement ``≈, <`` as basic, and define ``x ≤ y =  x < y ⊎ x ≈ y``
or
(B) implement ``≈, ≤`` as basic, and define ``x < y =  x ≤ y × ¬ x ≈ y``.
or
(C) implement them in some way specific to D
    (like, for example, Standard does with ℕ).

When you define something for a particular case, it is desirable to follow
some generic construction, if possible.
To my mind, the approach (A) is the most natural. Because

a) it is generic,

b) ``≤`` is often called "less or equal", so that it intends to express ``≤``
   as a disjunction of the two basics.

c) If one defines ``<`` via ``≤``, one still will have to prove
   ``x ≤ y ←→ x < y ⊎ x ≈ y`` (may be under some assuptions).  

d) the formula in (A) is simpler than the formula in (B).


Standard defines ``<`` for ℕ via ``≤`` and ``suc``.
But this is specific for ℕ: there are many TotalOrder domains having ``≈, ≤``
and without a sensible ``suc``.

As to me, I would rather define for ℕ
```
data _<_ ...
  where
  0<s : ∀ {n} → zero < suc n
  s<s : ∀ {m n} → m < n → suc m < suc n
...
m ≤ n =  m < n ⊎ m ≡ n 
```

In the sequel you write about x, y being not inferred from  x<y.
But in fact, they are inferred (see my replies below).

Reply:
Okay well if it doesn't result in any unsolved metavariables, I'm a little bit happier with it. Before discussing it further, please could we check this by removing any remaining explicitly given implicit arguments in the code. I've highlighted some of them in my latest review.